### PR TITLE
R tensorproduct

### PR DESCRIPTION
**Pull request by tkarna**
Adds support for foo x Real extruded function spaces.

### CODE REVIEW

**Comment by tkarna** [comment]
Requires firedrakeproject/tsfc#173

File: tests/extrusion/test_real_tensorproduct.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    mesh = extmesh(2, 2, 8)
+    fspace = FunctionSpace(mesh, hfamily, hdegree, vfamily=vfamily, vdegree=vdegree)
+
+    u = TrialFunction(fspace)
+    v = TestFunction(fspace)
+
+    xyz = SpatialCoordinate(mesh)
+    if functype == 'sin':
+        expr = sin(pi*xyz[2])
+        solution_expr = Constant(2.0/pi)
```
Comment (dham) [inline tests/extrusion/test_real_tensorproduct.py]: This is a bit icky. Why not just put the expression and solution in the pytest parametrisation?

Comment (tkarna) [reply tests/extrusion/test_real_tensorproduct.py]: Could do that too.

Comment (tkarna) [reply tests/extrusion/test_real_tensorproduct.py]: I can modify this to use string expressions if it's better.

Comment (wence-) [reply tests/extrusion/test_real_tensorproduct.py]: Use parametrised fixtures

```python
@pytest.fixture
def V(extmesh):
    mesh = extmesh(2, 2, 8)
    return FunctionSpace(mesh, "CG", 1, vfamily="Real", vdegree=1)

@pytest.fixture(params=["linear", "sin"]):
def variant(request):
    return request.param

@pytest.fixture
def expr(variant, V):
    x, y, z = SpatialCoordinate(V.ufl_domain())
    if variant == "linear":
        return z
    else:
       return sin(pi*z)

@pytest.fixture
def solution(variant):
    return {"linear": Constant(0.5), "sin": Constant(2.0/pi)}[variant]

@pytest.fixture
def tolerance(variant):
    return {"linear": 1e-15, "sin": 1e-5}[variant]

def test_foo(V, expr, solution, tolerance):
   ...
```

Comment (tkarna) [reply tests/extrusion/test_real_tensorproduct.py]: Huh, okay. Will do.

File: firedrake/mesh.py [hunk 2]
```diff
@@ -902,10 +906,11 @@ def make_offset(self, entity_dofs, ndofs):
             entity_offset[b] += len(entities[0])
 
         dof_offset = np.zeros(ndofs, dtype=IntType)
-        for (b, v), entities in entity_dofs.items():
-            for dof_indices in entities.values():
-                for i in dof_indices:
-                    dof_offset[i] = entity_offset[b]
```
Comment (wence-) [inline firedrake/mesh.py]: I wonder if it works to `return None` in the real case.  That is more "correct"

Comment (tkarna) [reply firedrake/mesh.py]: Doesn't seem to work out of the box.

Comment (wence-) [reply firedrake/mesh.py]: Might work in TJ's new reworking of the pyop2 codegen (I guess this is where things fail?).  Let's keep it in the back of our minds.

Comment (tkarna) [reply firedrake/mesh.py]: OK

**Comment by wence-** [comment]
I think this is fine. Will just wait for Jenkins to run through now that I've merged the TSFC branch.

**Pull request merged**